File: src/main/scala/pl/msulima/vistula/statments/FunctionDef.scala
package pl.msulima.vistula.statments

import pl.msulima.vistula.Ast
import pl.msulima.vistula.Ast.stmt

object FunctionDef {

  val apply: PartialFunction[stmt, String] = {
    case stmt.FunctionDef(name: Ast.identifier, args: Ast.arguments, body: Seq[stmt], _) =>
      val argumentNames = args.args.map(arg => arg match {
        case Ast.expr.Name(id, _) =>
          id.name
      }).mkString(", ")

      s"""
         |function ${name.name}($argumentNames) {
         |  Rx.Observable.zip($argumentNames, function($argumentNames) {
         |    return ${Statement(body.head)}
         |  }
         |};""".stripMargin
  }
}

File: src/test/resources/pl/msulima/vistula/transpiler/control/function_def.js
function a(X) {
    return X.rxMap($arg => ($arg + 2));
};
function b(X, Y) {
    const Z = vistula.zip([
        X,
        Y
    ]).rxMap($args => ($args[0] + $args[1]));
    return Z.rxMap($arg => ($arg + 2));
};